### DISCUSSION TITLE
Reimplement hash library using non-licensed, MIT or BSD code

### DISCUSSION THREAD

**Discussion by mcfriend99** [zuri-lang/zuri#115] [General]
A good catch by @galexite shows that the sources from which the hash functions in the `hash` module derived  seems to use unattributed code from the PHP source.

This all needs to be purged and replaced with new implementations or MIT/BSD licensed code.

**Comment by clsource** [comment]
A good alternative is https://www.libtom.net/LibTomCrypt/

**Comment by mcfriend99** [comment]
Thanks. I love it's license and I'll look through it for inspiration on replacing the current implementation. I'll also use this medium to solicit help from any body interested in helping with the transition of the hash methods. PRs for this issue are very welcomed.

**Comment by clsource** [comment]
As far as I know PHP License is not restrictive and is compatible with MIT

- https://www.php.net/license/3_01.txt
- https://www.php.net/license/index.php

It would be needed to include the license comment in the affected source code and put 

> "This product includes PHP software, freely available from
     <http://www.php.net/software/>".

somewhere

According to: https://joinup.ec.europa.eu/collection/eupl/matrix-eupl-compatible-open-source-licences

> Nothing in the license restricts the distribution of larger works in binary form under another license (as the EUPL).
Redistributions must retain the copyright notice, disclaimer and conditions (including restrictions on the use of “PHP” in the name of derived products).
These inherited conditions / restrictions are not “added” by the EUPL licensor and are in line with the EUPL requirements.

**Comment by mcfriend99** [comment]
I think it's counter productive for Blade MIT license. For those who plan on embedding Blade, this would mean they'll have to advertise The use of PHP in their product as well.

I believe the Unlicense and BSD license are very much more permissive and offer better interoperability with the MIT license.
